executor/coprocessor.go

```diff
@@ -0,0 +1,161 @@
+package executor
+
+import (
+	"context"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/coprocessor"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/timeutil"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+// HandleCopDAGRequest handles the coprocessor request.
+func HandleCopDAGRequest(ctx context.Context, sctx sessionctx.Context, req *coprocessor.Request) *coprocessor.Response {
+	resp := &coprocessor.Response{}
+	e, dagReq, err := buildDAGExecutor(sctx, req)
+	if err != nil {
+		resp.OtherError = err.Error()
+		return resp
+	}
+
+	err = e.Open(ctx)
+	if err != nil {
+		resp.OtherError = err.Error()
+		return resp
+	}
+
+	selResp := &tipb.SelectResponse{}
+	chk := newFirstChunk(e)
+	tps := e.base().retFieldTypes
+	for {
+		chk.Reset()
+		err = Next(ctx, e, chk)
+		if err != nil {
+			break
+		}
+		if chk.NumRows() == 0 {
+			break
+		}
+		err = fillUpData4SelectResponse(selResp, dagReq, sctx, chk, tps)
+		if err != nil {
+			break
+		}
+	}
+	return buildResp(selResp, err)
+}
+
+func buildDAGExecutor(sctx sessionctx.Context, req *coprocessor.Request) (Executor, *tipb.DAGRequest, error) {
+	if req.GetTp() != kv.ReqTypeDAG {
+		return nil, nil, errors.Errorf("unsupported request type %d", req.GetTp())
+	}
+	dagReq := new(tipb.DAGRequest)
+	err := proto.Unmarshal(req.Data, dagReq)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	sctx.GetSessionVars().StmtCtx.SetFlagsFromPBFlag(dagReq.Flags)
+	sctx.GetSessionVars().StmtCtx.TimeZone, err = timeutil.ConstructTimeZone(dagReq.TimeZoneName, int(dagReq.TimeZoneOffset))
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	e, err := buildDAG(sctx, dagReq.Executors)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	return e, dagReq, nil
+}
+
+func buildDAG(sctx sessionctx.Context, executors []*tipb.Executor) (Executor, error) {
+	var src core.PhysicalPlan
+	is := sctx.GetSessionVars().TxnCtx.InfoSchema.(infoschema.InfoSchema)
+	bp := core.NewPBPlanBuilder(sctx, is)
+	for i := 0; i < len(executors); i++ {
+		curr, err := bp.PBToPhysicalPlan(executors[i])
```

Comment:
Seems it's not a builder sematic. Prefer to move the `for` to the builder and the build function input a executors array and returns a physical plan.

Reply:
Great, done.